### ISSUE TITLE
V4l2loopback does not work as a video input

### ISSUE THREAD

**Issue by JulioSDDR** [jitsi/jitsi-meet#5186]
## Description
I use OBS with a v4l2 loopback to be able to 'chroma' my background while video-chating.  It is useful for me, it really gives me flexibility to share screen+video etc. I am unable to feed from the v4l2loopback to jitsi. The setup is working everywhere else, hangouts, googlemeet, skype web, and even webcamtests.com. Issue seems to match #1964 

## Current behavior
Jitsi refuses to accept the v4l2loopback input. It is not listed at the browser pop-up that asks for permissions. It can be listed in the settings within the chat-roomt, but it refuses to use it.  I can see it advertised in the browser jsconsole in:  [features/base/devices]

## Expected Behavior
Being able to choose v4l2loopback as a video device.

# Environment details
I am using Linux 5.5.8 with Firefox 73. It happens with other browsers (chromium) and linux versions.  
OBS 24 and v4l2loopback-dkms 0.12.3.
obs-v4l2sink as plugin.

## Steps to reproduce
install all the above
`modprobe v4l2loopback exclusive_caps=1`
obs»tools»v4l2sink start
go to a room meet.jit.si
the "dummy device" wont appear in the permission pop-up

**Comment by saghul** [comment]
Can you test it with test.webrtc.org ? What supported resolutions do you get?

**Comment by WolfgangFahl** [comment]
I am having the same issue with a CamTwist virtual camera on Mac OS 10.13.6 Testing with Chrome 80 on test.webrtc.org shows that everything is tested fine .eg. at resolutions 320x240, 640x480 and 1280x720.

**Comment by saghul** [comment]
Even for the CamTwist camera? I just tested it myself and it won't even show up. Likely because the driveris not 64bit compatible: http://camtwiststudio.com/support/faq/

**Comment by xqms** [comment]
I just used Jitsi with v4l2loopback and it worked fine. Video source in my case was ffmpeg, with pixel format yuv420p. Maybe that's important (it is for Skype at least).
`ffmpeg ...<input config>... -f v4l2 -filter:v "format=pix_fmts=yuv420p" /dev/videoX`

**Comment by partofthething** [comment]
Chromium/Chrome are very particular about enumerating devices for some reason. I have been able to get it to work though a somewhat convoluted series of steps:

* Close OBS and Chromium. Make sure nothing's using the main webcam
* Open Chromium and enter a video conference. Choose the main camera. 
* Now open OBS. Activate the [OBS V4L2 sink plugin](https://github.com/CatxFish/obs-v4l2sink) and stream it to the loopback
* Over in Chromium, go to device settings. Now the loopback option shows up! Switch to it. 
* Over in OBS, adjust video capture settings away from and then back to the main camera. This "steals" it back from Chromium

Now everything works. 

In Firefox it's much simpler, but there's that ongoing Firefox multicast issue so that doesn't help much.

**Comment by dromer** [comment]
For chromium you need to load the module with `exclusive_caps=1`

What I currently use:
`sudo modprobe v4l2loopback devices=1 video_nr=10 card_label="OBS Cam" exclusive_caps=1`

**Comment by DavidS** [comment]
I've been using the stack @JulioSDDR describes for a while now with great success using bluejeans in firefox and chrome. The trick for me has been to make sure that the loopback device is loaded first (and the USB devices (yes, two!) are loaded in a deterministic order).

The only way I could get jitsi to work was on chrome with the loopback device LAST and then choosing it in the settings after connecting to a room.

Notably, the Permission Request dialogs are different between bluejeans and jitsi, with the camera list not showing the loopback camera on firefox at all for jitsi, but fine for bluejeans.

tl;dr: browser handling of multiple cameras is broken, but there are video conferencing products out there that make it work despite it all.

![2020-04-27_11MS+0100_492x329](https://user-images.githubusercontent.com/27712/80362245-37b1ac80-887a-11ea-8b9b-36ba369f1d66.png)
![2020-04-27_11MS+0100_457x247](https://user-images.githubusercontent.com/27712/80362253-3a140680-887a-11ea-804c-592de6032987.png)

![2020-04-27_11MS+0100_882x897](https://user-images.githubusercontent.com/27712/80362247-384a4300-887a-11ea-88af-15902316950d.png)
![2020-04-27_11MS+0100_784x707](https://user-images.githubusercontent.com/27712/80362250-38e2d980-887a-11ea-8983-2627253ca976.png)
![2020-04-27_11MS+0100_888x575](https://user-images.githubusercontent.com/27712/80362251-397b7000-887a-11ea-9c37-ffacc5e87f1b.png)

**Comment by xqms** [comment]
While trying something else (an HDMI capture card with 1080p), I discovered that Jitsi Meet will hide devices that cannot supply 720p or below. Changing the device constraints in `config.js` will show these devices.
https://github.com/jitsi/jitsi-meet/blob/94a15914d037fa388aa022c7e8ede318368b8ae0/config.js#L113-L120

Specifically, you have to comment in the whole block and change `max` to `1080` (or whatever you need).

**Comment by ysalmon** [comment]
I am experiencing the same problem.

I usually set my virtual camera to 640x360 and Jitsi says it cannot access it "for an unknown reason" and it says access was not granted in the Settings (I cannot even enumerate the devices), although the webrtc test website detects it and acknowledges its resolution.

If I set the virtual camera to 1280x720, then Jitsi's main interface says the same, but this time the Settings dialog sees the device and gets its video feed !

Note that the (presumably older) Jitsi instance at https://rendez-vous.renater.fr/ uses the virtual camera correctly.

This is all under Chromium in KUbuntu 20.04.

**Comment by DavidS** [comment]
Using the instance at https://meet.jit.si/ chromium now picks up the virtual webcam automatically. Firefox still doesn't offer it to choose when asking for permission.

**Comment by saghul** [comment]
We've solved some devise issues by relaxing our resolution constraints. Closing.
